File: Stacks/Patterns/Pattern8/Pattern8.py
def next_greater(arr):
    n = len(arr)
    nge = [0] * n
    stack = []
    for i in range(n - 1, -1, -1):
        while stack and arr[stack[-1]] < arr[i]:
            stack.pop()
        if not stack:
            nge[i] = n
        else:
            nge[i] = stack[-1]
        stack.append(i)
    return nge


def prev_greater(arr):
    n = len(arr)
    pge = [0] * n
    stack = []
    for i in range(n):
        while stack and arr[stack[-1]] <= arr[i]:
            stack.pop()
        if not stack:
            pge[i] = -1
        else:
            pge[i] = stack[-1]
        stack.append(i)
    return pge


nums = [1, 3, 5]
nge = next_greater(nums)
pge = prev_greater(nums)
total = 0
for i in range(len(nums)):
    left = i - pge[i]
    right = nge[i] - i
    total += nums[i] * left * right
print("Sum of Subarray Maximums:", total)

File: Stacks/Patterns/Pattern8/Pattern8.java
import java.util.*;

public class Pattern8 {

    public static void main(String[] args) {
        int[] nums = {1, 3, 2};
        int[] nge = nextGreater(nums);
        int[] pge = prevGreater(nums);
        long sum = 0;
        for (int i = 0; i < nums.length; i++) {
            long left = i - pge[i];
            long right = nge[i] - i;
            sum += (long) nums[i] * left * right;
        }
        System.out.println(sum);
    }

    public static int[] nextGreater(int[] arr) {
        int n = arr.length;
        int[] nge = new int[n];
        Stack<Integer> st = new Stack<>();
        for (int i = n - 1; i >= 0; i--) {
            while (!st.isEmpty() && arr[st.peek()] < arr[i]) {
                st.pop();
            }
            if (st.isEmpty()) {
                nge[i] = n;
            } else {
                nge[i] = st.peek();
            }
            st.push(i);
        }
        return nge;
    }

    public static int[] prevGreater(int[] arr) {
        int n = arr.length;
        int[] pge = new int[n];
        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && arr[st.peek()] <= arr[i]) {
                st.pop();
            }
            if (st.isEmpty()) {
                pge[i] = -1;
            } else {
                pge[i] = st.peek();
            }
            st.push(i);
        }
        return pge;
    }
}
